From 9b5fa826ffa318ddf77ad88ef2e4a13c13fc6af6 Mon Sep 17 00:00:00 2001
From: Brian Belleville <bbellevi@uci.edu>
Date: Wed, 7 Mar 2018 16:02:45 -0800
Subject: [PATCH] floppy: Do not copy a kernel pointer to user memory in
 FDGETPRM ioctl

Backport from https://lkml.org/lkml/2018/3/7/1116

The final field of a floppy_struct is the field "name", which is a
pointer to a string in kernel memory. The kernel pointer should not be
copied to user memory. The FDGETPRM ioctl copies a floppy_struct to
user memory, including the "name" field. This pointer cannot be used
by the user, and it will leak a kernel address to user-space, which
will reveal the location of kernel code and data and undermine KASLR
protection. Instead, copy the floppy_struct except for the "name"
field.

Signed-off-by: Brian Belleville <bbellevi@uci.edu>
Signed-off-by: Jibin Xu <jibin.xu@windriver.com>
---
 drivers/block/floppy.c | 1 +
 1 file changed, 1 insertion(+)

diff --git a/drivers/block/floppy.c b/drivers/block/floppy.c
index 60d4c76..988ce70 100644
--- a/drivers/block/floppy.c
+++ b/drivers/block/floppy.c
@@ -3461,6 +3461,7 @@ static int fd_locked_ioctl(struct block_device *bdev, fmode_t mode, unsigned int
 					  (struct floppy_struct **)&outparam);
 		if (ret)
 			return ret;
+		size = offsetof(struct floppy_struct, name);
 		break;
 	case FDMSGON:
 		UDP->flags |= FTD_MSG;
-- 
1.8.5.2.233.g932f7e4

